heat/nn/data_parallel.py

```diff
@@ -0,0 +1,275 @@
+
+
+class DataParallel(tnn.Module):
+    """
+    Implements data parallelism across multiple processes. This means that the same model will be run locally
+    on each process. Creation of the model parallels to PyTorch, the only changes are using HeAT layers (ht.nn.layer)
+    in the initialization of the network. I.E. (example code contains)
+    .. code-block:: python
+
+        class TestingModel(torch.nn.Module):
+            def __init__(self):
+                super(TestingModel, self).__init__()
+                self.net1 = ht.nn.Linear(10, 10)
+                self.relu = ht.nn.ReLU()
+                self.net2 = ht.nn.Linear(10, 5)
+
+            def forward(self, x):
+                return self.net2(self.relu(self.net1(x)))
+
+        ht_model = ht.nn.DataParallel(TestingModel(), comm, optimizer)
+
+    and a requirement of giving a HeAT communicator (``comm``) as well as a Torch optimizer (``optimizer``). For the
+    given model both the ``__init__()`` and ``forward()`` functions must be defined in the class defining the network.
+
+    It is highly recommended that a HeAT DataLoader is used, see :func:`..utils.data.datatools.DataLoader`. The
+    default communications scheme for this is blocking. The blocking scheme will average the model parameters during
+    the backwards step, synchronizing them before the next model iteration.
+
+    To use the non-blocking communications for parameter updates, negate the optional flag ``blocking``.
+
+    Attributes
+    ----------
+    module : torch.nn.Module
+        the local module
+    comm : heat.MPICommunicator
+        HeAT communicator to use
+    optimizer : torch.optim.Optimizer
+        Optimizer used for parameter updates.
+    blocking : bool (optional)
+        Flag for blocking synchronization. If not given, synchronization is blocking by default.
+    """
+
+    def __init__(
+        self,
+        module: torch.nn.Module,
+        comm: ht.MPICommunication,
+        optimizer: torch.optim.Optimizer,
+        blocking: bool = True,
+    ):
+        super(DataParallel, self).__init__()
+        self.module = module
+        self.comm = comm
+        self.optimizer = optimizer
+        self.blocking = blocking
+
+        self._layer_wait_handles = OrderedDict()
+        self._fwd_hook_handles = list()
+        # set of layers' names with active wait handles (only relevant for non-blocking)
+        self._active_layers = set()
+        # slices of parameters belonging to one and the same layer
+        self._param_slices = dict()
+        # pytorch internal parameter indexing
+        self._param_indices = dict()
+        # reference of optimizer's params
+        self._params_ref = None
+        # flag indicating if optimizer should take a step during next iteration (only relevant for non-blocking)
+        self._update_next = False
+
+        # check if optimizer matches module
+        if list(module.parameters()) != optimizer.param_groups[0]["params"]:
```

Comment:
Sometimes I like to optimize only parts of the parameters, which seems to not possible here ?

Reply:
im not sure. do you have an example of how you would do that in a standard torch network? im unfamiliar with how the code would look for that

Reply:
@krajsek Could you please provide a small example of how you update only a part of the parameters?

Are you using it for transfer learning or what is the intention?

Reply:
I don't really want to do this now. It was rather meant to give the possibility to pass a subset. You can disable parameters with, requires_grad, but it is also allowed to pass only a part of the parameters, simple example: pass filter(lambda p: p.requires_grad, model.parameters()) instead of model.parameters() after some parameters are marked inactive

Reply:
For blocking communication, this is trivial to be implemented. For non-blocking, it'll be more complicated. But I’ll see what I can do.

Reply:
Could you not just gather all parameters in the optimizer at the beginning and then just work with them?

Reply:
Maybe I didn't understand your suggestion correctly, but I don't see how this could help. The problem with multiple optimizers and non-blocking parameter updates is that parameter updates are performed layer-wise currently. Since one optimizer could possibly deal with only a part of a layer's parameters and another optimizer could deal with another part, parameter updates would have to be performed parameter-wise to cover the general case. And this would increase not only code complexity but also computation time, because `optimizer.step()` wouldn't have to be invoked per layer as it is now, but per parameter. If we consider linear and conv layers, which contain two parameters each (bias and weight), this would result in doubling computation time caused by network's optimization step.

Reply:
What do you mean with multiple optimizers? Currently you can only pass one optimizer.

Reply:
Exactly, but we work on allowing to pass a bunch of optimizers. If one decides to use more than one though, communication during parameter updates will be blocking. This is the trade-off/compromise we have to accept.

Reply:
o.k. that is new for me. Can you give me an example where you need multiple optimizers?

Reply:
E.g. https://discuss.pytorch.org/t/different-optimisers-in-different-layers/72484
May be a corner case, but still.

Passing a single optimizer which deals only with a subset of parameters will be possible as well (this was your original request) and maybe will also work for non-blocking, but as I said, this is tricky. For blocking, it's trivial because optimizer.step() is invoked once to update all parameters being assigned. I don't have to care if it's actually a subset of parameters of if it's the total set of parameters. 
But for non-blocking, it's more complex. Here, parameters are updated layer-wise, so I need a mapping: layer -> parameter. This mapping isn't provided by parameters attribute of optimizer, therefore I have to do it on my own. And this becomes complex, if set of parameters being held by optimizer isn't equal to model's set of parameters.
Sorry for probably misleading in my former answer, I hope the actual problem of your original request is clear now.

Reply:
o.k., now it is clear for me, thanks. Maybe we can consider the  feature of having only a subset of parameters for later implementations